### ISSUE TITLE
recurrence of Cannot import local modules from notebook issue #43?

### ISSUE THREAD

**Issue by mrjeffstevenson3** [pauldmccarthy/fsleyes#80]
Hi Paul et al, just created a conda environment for python 3.8 and installed fsleyes and fslpy via conda-forge. I get the same error as Jorg in issue #43. Jupyter server fails.
`conda create --name py38fsl python=3.8`
`conda activate py38fsl`
`conda install -c conda-forge python=3.8 fsleyes fslpy`
`which fsleyes`
/Users/mrjeffs/Software/anaconda2/envs/py38fsl/bin/fsleyes
`fsleyes --version`
fsleyes/FSLeyes version 1.0.11
(sorry for the long traceback)
`fsleyes --notebookFile fsleyes_tutorial.ipynb`
   ERROR            status.py  161: reportIfError   - Error initialising IPython/Jupyter: An error occurred initialising IPython/Jupyter! 
 WARNING          __init__.py  695: create          - GLContext callback function raised RuntimeError: Could not start notebook server: Traceback (most recent call last):
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 22, in GL
    return ctypesloader.loadLibrary(
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/ctypesloader.py", line 36, in loadLibrary
    return _loadLibraryWindows(dllType, name, mode)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/ctypesloader.py", line 89, in _loadLibraryWindows
    return dllType( name, mode )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: ('dlopen(OSMesa, 10): image not found', 'OSMesa', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/main.py", line 910, in <module>
    main()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/main.py", line 235, in main
    fsleyes.initialise()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/__init__.py", line 243, in initialise
    import fsleyes.plugins as plugins
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/plugins/__init__.py", line 196, in <module>
    import fsleyes.views.canvaspanel     as canvaspanel
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/views/canvaspanel.py", line 20, in <module>
    from . import                    colourbarpanel
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/views/colourbarpanel.py", line 15, in <module>
    import fsleyes.gl.wxglcolourbarcanvas as cbarcanvas
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/gl/wxglcolourbarcanvas.py", line 17, in <module>
    import fsleyes.gl.colourbarcanvas as cbarcanvas
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/gl/colourbarcanvas.py", line 21, in <module>
    import OpenGL.GL as gl
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/GL/__init__.py", line 3, in <module>
    from OpenGL import error as _error
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/error.py", line 12, in <module>
    from OpenGL import platform, _configflags
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/__init__.py", line 36, in <module>
    _load()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/__init__.py", line 33, in _load
    plugin.install(globals())
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 97, in install
    namespace[ name ] = getattr(self,name,None)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 66, in GetCurrentContext
    function = self.OSMesa.OSMesaGetCurrentContext
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 60, in OSMesa
    def OSMesa( self ): return self.GL
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 28, in GL
    raise ImportError("Unable to load OpenGL library", *err.args)
ImportError: ('Unable to load OpenGL library', 'dlopen(OSMesa, 10): image not found', 'OSMesa', None)
Traceback (most recent call last):
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/gl/__init__.py", line 692, in create
    ready()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/main.py", line 574, in realCallback
    callback()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/main.py", line 396, in buildGui
    frame.menuActions[NotebookAction](namespace[0].notebookFile)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/actions/base.py", line 200, in __call__
    return self.__func(*args, **kwargs)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/actions/notebook.py", line 148, in __openNotebooks
    self.__server = self.__startServer(progdlg, nbfile)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/actions/notebook.py", line 208, in __startServer
    raise RuntimeError('Could not start notebook server: '
RuntimeError: Could not start notebook server: Traceback (most recent call last):
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 22, in GL
    return ctypesloader.loadLibrary(
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/ctypesloader.py", line 36, in loadLibrary
    return _loadLibraryWindows(dllType, name, mode)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/ctypesloader.py", line 89, in _loadLibraryWindows
    return dllType( name, mode )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: ('dlopen(OSMesa, 10): image not found', 'OSMesa', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/main.py", line 910, in <module>
    main()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/main.py", line 235, in main
    fsleyes.initialise()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/__init__.py", line 243, in initialise
    import fsleyes.plugins as plugins
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/plugins/__init__.py", line 196, in <module>
    import fsleyes.views.canvaspanel     as canvaspanel
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/views/canvaspanel.py", line 20, in <module>
    from . import                    colourbarpanel
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/views/colourbarpanel.py", line 15, in <module>
    import fsleyes.gl.wxglcolourbarcanvas as cbarcanvas
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/gl/wxglcolourbarcanvas.py", line 17, in <module>
    import fsleyes.gl.colourbarcanvas as cbarcanvas
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/gl/colourbarcanvas.py", line 21, in <module>
    import OpenGL.GL as gl
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/GL/__init__.py", line 3, in <module>
    from OpenGL import error as _error
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/error.py", line 12, in <module>
    from OpenGL import platform, _configflags
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/__init__.py", line 36, in <module>
    _load()
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/__init__.py", line 33, in _load
    plugin.install(globals())
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 97, in install
    namespace[ name ] = getattr(self,name,None)
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 66, in GetCurrentContext
    function = self.OSMesa.OSMesaGetCurrentContext
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 60, in OSMesa
    def OSMesa( self ): return self.GL
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mrjeffs/Software/anaconda2/envs/py38fsl/lib/python3.8/site-packages/OpenGL/platform/osmesa.py", line 28, in GL
    raise ImportError("Unable to load OpenGL library", *err.args)
ImportError: ('Unable to load OpenGL library', 'dlopen(OSMesa, 10): image not found', 'OSMesa', None)

any help getting a synchronized fsleyes notebook running would be greatly appreciated.

**Comment by pauldmccarthy** [comment]
Hi @mrjeffstevenson3 I don't have access to a mac at the moment (I normally have remote access to one, but the internal network is down at work), but I'll try and reproduce this when I can.

Does FSLeyes work normally, i.e. if you start it without the `--notebook` or `--notebookFile` options?

**Comment by mrjeffstevenson3** [comment]
yes works fine. but when I call File/Open Notebooks same error.
when I start from FSLeyes.app the notebook server starts but the kernel shows this error:
Traceback (most recent call last):
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/web.py", line 1704, in _execute
    result = await result
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 769, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/notebook/services/sessions/handlers.py", line 72, in post
    type=mtype))
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 769, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/notebook/services/sessions/sessionmanager.py", line 88, in create_session
    kernel_id = yield self.start_kernel_for_session(session_id, path, name, type, kernel_name)
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 769, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/notebook/services/sessions/sessionmanager.py", line 101, in start_kernel_for_session
    self.kernel_manager.start_kernel(path=kernel_path, kernel_name=kernel_name)
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/tornado/gen.py", line 216, in wrapper
    result = ctx_run(func, *args, **kwargs)
  File "/var/folders/77/rw4g32k56w31nfw8q9hw303r0000gn/T/fsleyes-jupyter42emcfb6/config/fsleyes_kernelmanager.py", line 49, in start_kernel
    .start_kernel(**kwargs).result()
AttributeError: 'coroutine' object has no attribute 'result'

Can't seem to get a running notebook with working kernel from any method.

**Comment by pauldmccarthy** [comment]
The second error is entirely unrelated to the first, and is probably due to a mismatch between the version of FSLeyes and the version of tornado - this should have been fixed in FSLeyes [here](https://github.com/pauldmccarthy/fsleyes/issues/62).

You are actually using two independent installations of FSLeyes - FSLeyes.app is coming from a FSL installation, at `/usr/local/fsl/`. So perhaps the `coroutine` error is arising from a previous failed update.  You could try and fix _that_ installation by running `$FSLDIR/fslpython/bin/conda update -n fslpython -c conda-forge --update-all fsleyes`

The original error (`Unable to load OpenGL library`) is due to the way that the conda-forge `python.app` package works around restrictions on macOS. I can code up a workaround in FSLeyes though - will aim to release a new version this week. In the meantime, as a temporary measure to get your installation working, you could edit `~/anaconda2/envs/py38fsl/lib/python3.8/site-packages/fsleyes/actions/notebook.py`, and change line 603 from:

```
cmd = [sys.executable, fsleyes_main.__file__]
```

to:

```
cmd = [op.join(sys.exec_prefix, 'bin', 'pythonw'), fsleyes_main.__file__]
```

**Comment by mrjeffstevenson3** [comment]
thanks @pauldmccarthy the new cmd at the end fixed the py38fsl anaconda env method. unfortunately the conda update  to  /usr/local/fsl/ broke the notebook server launch altogether (kernel wasn't working anyway so no great loss) in case it helps. 

 `which fsleyes`
/usr/local/fsl/bin/fsleyes
FSLeyes version 1.0.11
an incomplete traceback due to bad scrolling:

An error occurred initialising IPython/Jupyter! 

Details: Could not start notebook server: Traceback (most recent call last):
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 25, in GL
    mode=ctypes.RTLD_GLOBAL 
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 36, in loadLibrary
    return _loadLibraryWindows(dllType, name, mode)
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 89, in _loadLibraryWindows
    return dllType( name, mode )
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/ctypes/__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: ('dlopen(OSMesa, 10): image not found', 'OSMesa', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/main.py", line 910, in <module>
    main()
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/main.py", line 235, in main
    fsleyes.initialise()
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/__init__.py", line 243, in initialise
    import fsleyes.plugins as plugins
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/plugins/__init__.py", line 196, in <module>
    import fsleyes.views.canvaspanel     as canvaspanel
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/views/canvaspanel.py", line 20, in <module>
    from . import                    colourbarpanel
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/views/colourbarpanel.py", line 15, in <module>
    import fsleyes.gl.wxglcolourbarcanvas as cbarcanvas
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/gl/wxglcolourbarcanvas.py", line 17, in <module>
    import fsleyes.gl.colourbarcanvas as cbarcanvas
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/gl/colourbarcanvas.py", line 21, in <module>
    import OpenGL.GL as gl
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/GL/__init__.py", line 3, in <module>
    from OpenGL import error as _error
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/error.py", line 12, in <module>
    from OpenGL import platform, _configflags
  File "/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/pla

**Comment by pauldmccarthy** [comment]
I'm guessing it's the same error as in your `py38fsl` environment - have you tried patching `notebook.py` the same as before? The file to edit would be `/usr/local/fsl/fslpython/envs/fslpython/lib/python3.7/site-packagesfsleyes/actions/notebook.py`

In any case, I've released version 1.0.12, which should be available in a few hours time, and should resolve this problem.

**Comment by mrjeffstevenson3** [comment]
hi @pauldmccarthy, both methods now working. awesome job!
